stdlib/seq.mli

```diff
@@ -13,97 +13,807 @@
+    @since 4.14 *)
+
+val drop : int -> 'a t -> 'a t
+(** [drop n xs] is the sequence [xs], deprived of its first [n] elements.
+
+    If [xs] has fewer than [n] elements,
+    then [drop n xs] is empty.
+
+    [n] must be nonnegative.
+
+    [drop] is lazy: the first [n+1] elements of the sequence [xs]
+    are demanded only when the first element of [drop n xs] is
+    demanded. For this reason, [drop 1 xs] is {i not} equivalent
+    to [tail xs], which queries [xs] immediately.
+
+    @raise Invalid_argument if [n] is negative.
+    @since 4.14 *)
+
+val take_while : ('a -> bool) -> 'a t -> 'a t
+(** [take_while p xs] is the longest prefix of the sequence [xs]
+    where every element [x] satisfies [p x].
+
+    @since 4.14 *)
+
+val drop_while : ('a -> bool) -> 'a t -> 'a t
+(** [drop_while p xs] is the sequence [xs], deprived of the prefix
+    [take_while p xs].
+
+    @since 4.14 *)
+
+val uniq : ('a -> 'a -> bool) -> 'a t -> 'a t
+(** Provided the function [eq] defines an equality on elements,
+    [uniq eq xs] is the sequence [xs],
+    deprived of its adjacent duplicate elements.
+
+    Thus, [uniq eq xs] has no adjacent duplicate elements.
+    Furthermore, if [xs] is sorted according to an ordering
+    that is compatible with [eq],
+    then [uniq eq xs] has no duplicate elements at all.
+
+    [uniq eq xs] is equivalent to [map head (group eq xs)].
+
+    @since 4.14 *)
+
+val group : ('a -> 'a -> bool) -> 'a t -> 'a t t
+(** Provided the function [eq] defines an equality on elements,
+    [group eq xs] is the sequence of the maximal runs
+    of adjacent duplicate elements of the sequence [xs].
+
+    Every element of [group eq xs] is a nonempty sequence of equal elements.
+
+    The concatenation [concat (group eq xs)] is equal to [xs].
+
+    Consuming [group eq xs], and consuming the sequences that it contains,
+    can cause [xs] to be consumed more than once. Therefore, [xs] must be
+    persistent.
+
+    @since 4.14 *)
+
+val memoize : 'a t -> 'a t
+(** The sequence [memoize xs] has the same elements as the sequence [xs].
+
+    Regardless of whether [xs] is ephemeral or persistent,
+    [memoize xs] is persistent: even if it is queried several times,
+    [xs] is queried at most once.
+    @since 4.14 *)
+
+exception Forced_twice
+(** This exception is raised when a sequence returned by [once]
+    (or a suffix of it) is queried more than once.
+    @since 4.14 *)
+
+val once: 'a t -> 'a t
+(** The sequence [once xs] has the same elements as the sequence [xs].
+
+    Regardless of whether [xs] is ephemeral or persistent,
+    [once xs] is an ephemeral sequence: it can be queried at most once.
+    If it (or a suffix of it) is queried more than once, then the exception
+    [Forced_twice] is raised. This can be useful, while debugging or testing,
+    to ensure that a sequence is consumed at most once.
+
+    @raise Forced_twice if [once xs], or a suffix of it,
+           is queried more than once.
+    @since 4.14 *)
```

Comment:
Musing question: is there a more general function than `memoize` and `once` that applies "thunk operators" to elements of a stream?

```
val map_thunks : ((unit -> 'a node) -> (unit -> 'a node)) -> 'a t -> 'a t
```

We could in fact consider defining `memoize` and `once`, on the type `(unit -> 'a)`, in the `Fun` module. (It would even make sense to expose `Suspension` as a submodule there.). I think that having `memoize` redefined here as `map_thunks Fun.Suspension.memoize` would be nice, but `Seq.once` might then not feel so necessary anymore. 

(And `map_thunks` has other uses, for example to compute the maximum latency of computing each node of the sequence.)